p2p/src/peer.rs

```diff
@@ -141,22 +163,28 @@ where
 }
 
 /// This is an endpoint, that juggles messages between [`crate::Network`] and another connected node.
-pub struct Peer<T, K, E>
+pub enum Peer<T, K, E>
 where
-    T: Debug + Encode + Decode + Send + Clone + 'static,
+    T: Encode + Decode + Send + Clone + 'static,
     K: KeyExchangeScheme + Send + 'static,
     E: Encryptor + Send + 'static,
 {
-    /// Peer identification
-    id: PeerId,
-    /// Peer connnection
-    connection: Connection,
-    /// Current peer/connection state
-    state: State,
-    /// Cryptographic data
-    crypto: Cryptographer<T, K, E>,
-    /// The to send received messages upstairs
-    broker: Broker,
+    /// Peer just created and beginning handshake process.
+    Connecting(PeerId, Broker),
+    /// Peer has a living TCP connection, ready to handshake.
+    ConnectedTo(PeerId, Broker, Connection),
+    /// Peer has just been connected to  from the outside.
+    ConnectedFrom(PeerId, Broker, Connection),
+    /// Peer ready to send its public key from its PeerId.
+    SendKey(PeerId, Broker, Connection, Cryptographer<T, K, E>),
+    /// Peer ready to read public key for PeerId.
+    GetKey(PeerId, Broker, Connection, Cryptographer<T, K, E>),
+    /// Peer completed handshake. Only this peer variant can send messages.
+    Ready(PeerId, Broker, Connection, Cryptographer<T, K, E>),
+    /// Peer has been (gracefully) disconnected.
+    Disconnected(PeerId),
```

Comment:
Thank you for showing me another implementation.
I think the previous version (your choice) is better